### PR TITLE
DEP Use PHPUnit 11

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/.github/issues/305

[PHPUnit 10](https://phpunit.de/announcements/phpunit-10.html) removed the ability to unit test Error's, Deprecation's, Warning's and Strict's, so I've set a custom error handler to convert all those to unit test specific exceptions instead

I've also updated a couple of dependencies which PHPUnit 11 requires:
- nikic/php-parser
- sebastian/diff

### CODE REVIEW

File: src/Dev/SSListExporter.php
Comment (emteknetnz) [inline src/Dev/SSListExporter.php:1]: Exporter is now a final class, so this class must be removed

This lose in functionality will make it a bit harder to see why a unit-test failed when using the following assertions:
- assertListContains (used a handful of times)
- assertListNotContains (hardly used)
- assertListEquals (used quite a lot)
- assertListAllMatch (hardly used)


The SS_List will get displayed however https://github.com/sebastianbergmann/exporter/blob/main/src/Exporter.php wants it to now

File: tests/php/Dev/SSListExporterTest.php
Comment (emteknetnz) [inline tests/php/Dev/SSListExporterTest.php:1]: Being removed as SSListExporter is being removed

File: src/Dev/SapphireTest.php
Comment (emteknetnz) [inline src/Dev/SapphireTest.php:1221]: `TestUtil::parseTestMethodAnnotations` was an `@internal` phpunit method that was removed without a deprecation warning. I've reimplemented it here and added a unit test for it in SapphireTestTest.php

File: tests/php/Control/DirectorTest.php
Comment (emteknetnz) [inline tests/php/Control/DirectorTest.php:764]: runTest() is now a final method so it cannot be overriden. I've reimplemented this functionality by changing it the testing methodology to `withExpectedExceptionRedirect()` instead

File: src/Dev/Constraint/SSListContains.php
Comment (emteknetnz) [inline src/Dev/Constraint/SSListContains.php:34]: SSListExporter has been removed, so removing it here.

File: src/Dev/Constraint/SSListContainsOnlyMatchingItems.php
Comment (emteknetnz) [inline src/Dev/Constraint/SSListContainsOnlyMatchingItems.php:31]: SSListExporter has been removed, so removing it here.

File: src/Dev/SSListExporter.php
Comment (emteknetnz) [reply src/Dev/SSListExporter.php:1]: Seems like we're not loosing anything

I tested by modifying the expected value in GridFieldDetailFormTest::testEditForm():239

Previously a failure in assertListContains() would look like this:

```
Failed asserting that SilverStripe\ORM\HasManyList Object &0000000000000dc40000000000000000 (
    'dynamicData' => Array &0 ()
    'failover' => null
    'customisedObject' => null
    'objCache' => Array &1 ()
    'extension_instances' => null
    'beforeExtendCallbacks' => Array &2 ()
    'afterExtendCallbacks' => Array &3 ()
    'extra_method_registers' => Array &4 ()
    'dataClass' => 'SilverStripe\Forms\Tests\GridField\GridFieldDetailFormTest\Person'
    'dataQuery' => SilverStripe\ORM\DataQuery Object &0000000000000dc00000000000000000 (
        'dataClass' => 'SilverStripe\Forms\Tests\GridField\GridFieldDetailFormTest\Person'
        'query' => SilverStripe\ORM\Queries\SQLSelect Object &0000000000000dc30000000000000000 (
            'replacementsOld' => Array &5 ()
            'replacementsNew' => Array &6 ()
            'where' => Array &7 (
                0 => Array &8 (
                    '"GridFieldDetailFormTest_Person"."GroupID" = ?' => Array &9 (
                        0 => 1
                    )
                )
            )
            'connective' => 'AND'
            'from' => Array &10 (
                'GridFieldDetailFormTest_Person' => '"GridFieldDetailFormTest_Person"'
            )
            'select' => Array &11 ()
            'groupby' => Array &12 ()
            'having' => Array &13 ()
            'union' => Array &14 ()
            'with' => Array &15 ()
            'distinct' => true
            'orderby' => Array &16 (
                '"FirstName"' => 'ASC'
            )
            'limit' => Array &17 ()
        )
        'collidingFields' => Array &18 ()
        'allowCollidingFieldStatements' => false
        'dataQueryManipulators' => Array &19 ()
        'queriedColumns' => null
        'queryFinalised' => false
        'querySubclasses' => true
        'filterByClassName' => true
        'queryParams' => Array &20 (
            '' => null
            'Foreign.ID' => 1
            'Foreign.Filter' => Array &21 (
                '"GridFieldDetailFormTest_Person"."GroupID"' => 1
            )
        )
        'extension_instances' => null
        'beforeExtendCallbacks' => Array &22 ()
        'afterExtendCallbacks' => Array &23 ()
        'extra_method_registers' => Array &24 ()
    )
    'finalisedQuery' => null
    'eagerLoadRelationChains' => Array &25 ()
    'eagerLoadAllRelations' => Array &26 ()
    'eagerLoadedData' => Array &27 ()
    'inAlterDataQueryCall' => false
    'addCallbacks' => null
    'removeCallbacks' => null
    'foreignKey' => 'GroupID'
)  contains an item matching ( "Surname" is "Bagginsx").
```

Now it looks like this
```
Failed asserting that SilverStripe\ORM\HasManyList Object #2006 (
    'dynamicData' => Array &0 [],
    'failover' => null,
    'customisedObject' => null,
    'objCache' => Array &1 [],
    'extension_instances' => null,
    'beforeExtendCallbacks' => Array &2 [],
    'afterExtendCallbacks' => Array &3 [],
    'extra_method_registers' => Array &4 [],
    'dataClass' => 'SilverStripe\Forms\Tests\GridField\GridFieldDetailFormTest\Person',
    'dataQuery' => SilverStripe\ORM\DataQuery Object #2016 (
        'dataClass' => 'SilverStripe\Forms\Tests\GridField\GridFieldDetailFormTest\Person',
        'query' => SilverStripe\ORM\Queries\SQLSelect Object #2005 (
            'replacementsOld' => Array &5 [],
            'replacementsNew' => Array &6 [],
            'where' => Array &7 [
                0 => Array &8 [
                    '"GridFieldDetailFormTest_Person"."GroupID" = ?' => Array &9 [
                        0 => 1,
                    ],
                ],
            ],
            'connective' => 'AND',
            'from' => Array &10 [
                'GridFieldDetailFormTest_Person' => '"GridFieldDetailFormTest_Person"',
            ],
            'select' => Array &11 [],
            'groupby' => Array &12 [],
            'having' => Array &13 [],
            'union' => Array &14 [],
            'with' => Array &15 [],
            'distinct' => true,
            'orderby' => Array &16 [
                '"FirstName"' => 'ASC',
            ],
            'limit' => Array &17 [],
        ),
        'collidingFields' => Array &18 [],
        'allowCollidingFieldStatements' => false,
        'dataQueryManipulators' => Array &19 [],
        'queriedColumns' => null,
        'queryFinalised' => false,
        'querySubclasses' => true,
        'filterByClassName' => true,
        'queryParams' => Array &20 [
            '' => null,
            'Foreign.ID' => 1,
            'Foreign.Filter' => Array &21 [
                '"GridFieldDetailFormTest_Person"."GroupID"' => 1,
            ],
        ],
        'extension_instances' => null,
        'beforeExtendCallbacks' => Array &22 [],
        'afterExtendCallbacks' => Array &23 [],
        'extra_method_registers' => Array &24 [],
    ),
    'finalisedQuery' => null,
    'eagerLoadRelationChains' => Array &25 [],
    'eagerLoadAllRelations' => Array &26 [],
    'eagerLoadedData' => Array &27 [],
    'inAlterDataQueryCall' => false,
    'addCallbacks' => null,
    'removeCallbacks' => null,
    'foreignKey' => 'GroupID',
)  contains an item matching ( "Surname" is "Bagginsx").
```

File: src/Dev/FunctionalTest.php
Comment (emteknetnz) [inline src/Dev/FunctionalTest.php:137]: As far as I can tell `@uses` was for code-coverage

I'm not sure what the replacement for `@uses` is https://github.com/sebastianbergmann/phpunit/issues/4502

This was already not-recommended as `@uses ClassName::methodName (not recommended because too fine-grained)`

We're not doing any sort of code-coverage at this point

I did try using `#[UsesMethod('Director', 'test')]` however it will complain `Attribute "PHPUnit\Framework\Attributes\UsesMethod" cannot target method (allowed targets: class)`

I've elected to just remove these instead

File: tests/php/View/SSViewerTest.php
Comment (emteknetnz) [inline tests/php/View/SSViewerTest.php:2374]: The addMethods() call now throws a deprecation warning, will be removed in phpunit12, so have rewritten test

File: src/View/Parsers/HtmlDiff.php
Comment (emteknetnz) [inline src/View/Parsers/HtmlDiff.php:289]: sebastian/diff had a major bump from 4 to 6 as it's a requirement of phpunit, the signature for the constructor changed

File: tests/php/ORM/DBSchemaManagerTest.php
Comment (emteknetnz) [inline tests/php/ORM/DBSchemaManagerTest.php:75]: getMockForAbstractClass() is deprecated https://github.com/sebastianbergmann/phpunit/issues/5241

I'm not sure why this test is worried about things being overridden by child classes as we're getting a mock of the base class using phpunit, not Silverstripe Injector

I've simply switch this to use ->getMock() instead

File: src/Core/Manifest/ClassManifest.php
Comment (emteknetnz) [inline src/Core/Manifest/ClassManifest.php:305]: API change from new version on nikic/php-parser

File: tests/php/Security/Member_GroupSetTest.php
Comment (emteknetnz) [inline tests/php/Security/Member_GroupSetTest.php:37]: No idea why though this unit test was failing, almost like assertEqualsCanonicalizing() just wasn't doing what it was supposed to. I just sorted things the manual way and the test passes.

File: src/Dev/SapphireTest.php
Comment (GuySartorelli) [inline src/Dev/SapphireTest.php:1283]: I would have expected `nikic/php-parser` to be able to do this? Can we leverage that package to handle this for us instead?

File: tests/php/Control/HTTPRequestTest.php
Comment (GuySartorelli) [inline tests/php/Control/HTTPRequestTest.php]: This is a bad pattern, because it means any time we want to `expectException` we have to do it in this decentralised way. It would be easy to do this in 20 different ways across 20 different tests and means reimplementing or at least copying the logic every time we do it.

Implement a `expectWarning` method in `SapphireTest` instead, to mimic the old API that `TestCase` used to have.

Comment (GuySartorelli) [reply tests/php/Control/HTTPRequestTest.php]: Same with any other similar changes (probably there's `expectError` and `expectDeprecation` somewhere as well)

File: tests/php/Core/ObjectTest.php
Comment (GuySartorelli) [inline tests/php/Core/ObjectTest.php:16]: Why were these renamed?

File: tests/php/ORM/DBSchemaManagerTest.php
Comment (GuySartorelli) [inline tests/php/ORM/DBSchemaManagerTest.php]: Put the test class in its own file

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Mostly looks good, few changes requested.

I won't look at the other module PRs just yet as they may need to change based on the changes I've requested.

File: tests/php/Core/ObjectTest.php
Comment (emteknetnz) [reply tests/php/Core/ObjectTest.php:16]: Because they end in the word Test.

PHPUnit now automatically thinks anything ending in the word "Test" is a test

File: src/Dev/SapphireTest.php
Comment (emteknetnz) [reply src/Dev/SapphireTest.php:1283]: As far as a can tell php-parser will also just return the DocComment - https://github.com/nikic/PHP-Parser/blob/master/lib/PhpParser/Comment/Doc.php - it won't get any more finely grained than that and extract annotations for us.

Given that PHP natively has a getDocComment() method available, seems like there's no benefit to using php-parser for this scenario

File: tests/php/Control/HTTPRequestTest.php
Comment (emteknetnz) [reply tests/php/Control/HTTPRequestTest.php]: Updated

File: tests/php/ORM/DBSchemaManagerTest.php
Comment (emteknetnz) [reply tests/php/ORM/DBSchemaManagerTest.php]: Done

File: tests/php/ORM/DataQueryTest.php
Comment (emteknetnz) [inline tests/php/ORM/DataQueryTest.php:238]: While this previously had expectError(), running this code without $this->enableErrorHandler() will now throw a php warning, so expecting an ExepectedWarningException is correct

File: src/Dev/SapphireTest.php
Comment (GuySartorelli) [reply src/Dev/SapphireTest.php:1283]: That's a shame but fair enough. Thanks for taking a look.

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge once all other PRs are approved

**Review by GuySartorelli — APPROVED** [review]
Changes approved

**Pull request merged**
